packages/common/src/api/tan-query/developerApps.ts

```diff
@@ -0,0 +1,215 @@
+import { ID } from '~/models/Identifiers'
+
+import { QUERY_KEYS } from './queryKeys'
+import { QueryOptions } from './types'
+import { useCurrentUserId } from './useCurrentUserId'
+
+export const DEVELOPER_APP_DESCRIPTION_MAX_LENGTH = 128
+export const DEVELOPER_APP_NAME_MAX_LENGTH = 50
+export const DEVELOPER_APP_IMAGE_URL_MAX_LENGTH = 2000
+const DEVELOPER_APP_IMAGE_URL_REGEX = /^(https?):\/\//i
+
+const messages = {
+  invalidUrl: 'Invalid URL'
+}
+
+export const developerAppSchema = z.object({
+  userId: z.number(),
+  name: z.string().max(DEVELOPER_APP_NAME_MAX_LENGTH),
+  imageUrl: z.optional(
+    z
+      .string()
+      .max(DEVELOPER_APP_IMAGE_URL_MAX_LENGTH)
+      .refine((value) => DEVELOPER_APP_IMAGE_URL_REGEX.test(value), {
+        message: messages.invalidUrl
+      })
+  ),
+  description: z.string().max(DEVELOPER_APP_DESCRIPTION_MAX_LENGTH).optional()
+})
+
+export const developerAppEditSchema = z.object({
+  userId: z.number(),
+  apiKey: z.string(),
+  name: z.string().max(DEVELOPER_APP_NAME_MAX_LENGTH),
+  imageUrl: z.optional(
+    z
+      .string()
+      .max(DEVELOPER_APP_IMAGE_URL_MAX_LENGTH)
+      .refine((value) => DEVELOPER_APP_IMAGE_URL_REGEX.test(value), {
+        message: messages.invalidUrl
+      })
+  ),
+  description: z.string().max(DEVELOPER_APP_DESCRIPTION_MAX_LENGTH).optional()
+})
+
+export type DeveloperApp = {
+  name: string
+  description?: string
+  imageUrl?: string
+  apiKey: string
+  apiSecret?: string
+}
+
+type UseAddDeveloperAppArgs = Omit<DeveloperApp, 'apiKey'> & {
+  userId: number
+}
+
+type UseEditDeveloperAppArgs = Omit<DeveloperApp, 'apiSecret'> & { userId: ID }
+
+type UseDeleteDeveloperAppArgs = {
+  apiKey: string
+  userId: ID
+}
+
+export const useDeveloperApps = (options?: QueryOptions) => {
+  const { audiusSdk } = useAudiusQueryContext()
+  const { data: userId } = useCurrentUserId()
+
+  return useQuery({
+    queryKey: [QUERY_KEYS.authorizedApps, userId],
+    queryFn: async () => {
+      const sdk = await audiusSdk()
+      const { data = [] } = await sdk.users.getDeveloperApps({
+        id: Id.parse(userId)
+      })
+
+      return data
+    },
+    staleTime: options?.staleTime,
+    enabled: options?.enabled !== false && !!userId
+  })
+}
+
+export const useAddDeveloperApp = () => {
+  const { audiusSdk } = useAudiusQueryContext()
+
+  return useMutation({
+    mutationFn: async (args: UseAddDeveloperAppArgs) => {
+      const { name, description, imageUrl, userId } = args
+      const encodedUserId = Id.parse(userId)
+      const sdk = await audiusSdk()
+
+      const { apiKey, apiSecret } = await sdk.developerApps.createDeveloperApp({
+        name,
+        description,
+        imageUrl,
+        userId: encodedUserId
+      })
+
+      await sdk.grants.createGrant({
+        userId: encodedUserId,
+        appApiKey: apiKey
+      })
+
+      return { name, description, imageUrl, apiKey, apiSecret }
+    }
```

Comment:
assuming we should do some optimistic updates?

Reply:
i agree, made a note in the description. Not sure if it is fine to add one without the apiKey present from the creation so wanted you all's opinion on how to handle

Reply:
also it seems like it populates quickly enough that the user sees the new app right away so it seems like it's fine